File: 2021/day8/day8.rs
// Day 8: Seven Segment Search
// https://adventofcode.com/2021/day/8

use std::collections::HashMap;
use std::collections::HashSet;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use structopt::StructOpt;

/// parse command line arguments
#[derive(StructOpt)]
struct Cli {
    #[structopt(default_value = "input.txt", parse(from_os_str))]
    path: std::path::PathBuf,
}

/// main function
fn main() {
    let args = Cli::from_args();
    println!("reading data from: {}", args.path.display());
    let data = load_data(args.path);

    part1(&data);
    part2(&data);
}

fn contains(a: &str, b: &str) -> bool {
    for c in b.chars() {
        if !a.contains(c) {
            return false;
        }
    }
    true
}

fn part2(data: &[String]) {
    // Nota: the Python version is much more readable...
    let mut total = 0;
    for line in data {
        let (notes1, code1) = line.split_once('|').unwrap();

        let notes = notes1.split_whitespace().collect::<Vec<&str>>();
        let code = code1.split_whitespace().collect::<Vec<&str>>();

        let mut d: HashMap<usize, HashSet<String>> = HashMap::new();

        for c in notes {
            let mut s: Vec<char> = c.chars().collect();
            s.sort_unstable();
            let s_sorted = String::from_iter(s);
            d.entry(c.len())
                .or_insert_with(HashSet::new)
                .insert(s_sorted);
        }

        let mut map: HashMap<String, u8> = HashMap::new();

        let mut zero: &str = "";
        let one = d.get(&2).unwrap().iter().next().unwrap(); // the only digit with the 2 segments
        let mut two: &str = "";
        let mut three: &str = "";
        let four = d.get(&4).unwrap().iter().next().unwrap(); // the only digit with the 4 segments
        let mut five: &str = "";
        let mut six: &str = "";
        let seven = d.get(&3).unwrap().iter().next().unwrap(); // the only digit with the 3 segments
        let eight = d.get(&7).unwrap().iter().next().unwrap(); // the only digit with 7 segments
        let mut nine: &str = "";

        for x in d.get(&6).unwrap().iter() {
            // could be 0 6 9
            if contains(x, one) {
                // si on n'a pas les deux segments du 1
                if contains(x, four) {
                    // si 4 est inclus, c'est forcément un 9
                    nine = x;
                } else {
                    // sinon, c'est forcément un 0
                    zero = x;
                }
            } else {
                // sinon, c'est forcément un 6
                six = x;
            }
        }

        for x in d.get(&5).unwrap().iter() {
            // 2 3 5
            if contains(x, seven) {
                // si 7 est inclus, c est forcément un 3
                three = x;
            } else if contains(nine, x) {
                // si c'est inclus dans 9, c'est forcément un 5
                five = x;
            } else {
                // sinon, c'est forcément un 2
                two = x;
            }
        }

        map.insert(zero.to_string(), 0);
        map.insert(one.to_string(), 1);
        map.insert(two.to_string(), 2);
        map.insert(three.to_string(), 3);
        map.insert(four.to_string(), 4);
        map.insert(five.to_string(), 5);
        map.insert(six.to_string(), 6);
        map.insert(seven.to_string(), 7);
        map.insert(eight.to_string(), 8);
        map.insert(nine.to_string(), 9);

        let mut r = 0;
        for d in code {
            let mut s: Vec<char> = d.chars().collect();
            s.sort_unstable();
            let s_sorted = String::from_iter(s);

            let v = u32::from(*map.entry(s_sorted).or_insert(0));
            r = r * 10 + v;
        }
        total += r;
    }
    println!("{:?}", total);
}

fn part1(data: &[String]) {
    let mut digit_one = 0;
    let mut digit_four = 0;
    let mut digit_seven = 0;
    let mut digit_eight = 0;

    for line in data {
        let x = line.split_once('|').unwrap().1.trim();
        for c in x.split_whitespace() {
            match c.len() {
                2 => digit_one += 1,
                3 => digit_seven += 1,
                4 => digit_four += 1,
                7 => digit_eight += 1,
                _ => (),
            }
        }
    }

    println!("{}", digit_one + digit_four + digit_seven + digit_eight);
}

// The output is wrapped in a Result to allow matching on errors
// Returns an Iterator to the Reader of the lines of the file.
fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

/// load data from file
fn load_data(path: std::path::PathBuf) -> Vec<String> {
    let mut data = vec![];
    if let Ok(lines) = read_lines(path) {
        for line in lines.flatten() {
            data.push(line);
        }
    }
    data
}

#[cfg(test)]
#[test]
fn test_contains() {
    assert_eq!(contains("abcde", "ab"), true);
    assert_eq!(contains("abcde", "ba"), true);
    assert_eq!(contains("abcde", "abcde"), true);
    assert_eq!(contains("abcde", "az"), false);
    assert_eq!(contains("abcde", "ef"), false);
}

File: 2021/day8/day8.py
#!/usr/bin/env python3

# Day 8: Seven Segment Search
# https://adventofcode.com/2021/day/8

from collections import defaultdict
import sys

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = open(filename).readlines()


# Part 1
digit_one = 0
digit_four = 0
digit_seven = 0
digit_eight = 0

for line in data:
    x = line.split("|", maxsplit=1)[1].strip()
    for c in x.split():
        if len(c) == 2:
            digit_one += 1
        elif len(c) == 3:
            digit_seven += 1
        elif len(c) == 4:
            digit_four += 1
        if len(c) == 7:
            digit_eight += 1

print(digit_one + digit_four + digit_seven + digit_eight)

# part 2
total = 0
for line in data:
    notes, code = line.split("|", maxsplit=1)

    notes = notes.strip().split()
    code = code.strip().split()

    d = defaultdict(lambda: set())

    for c in notes:
        d[len(c)].add("".join(sorted(c)))

    # 2 segments : 1
    # 3 segments : 7
    # 4 segments : 4
    # 5 segments : 2 3 5
    # 6 segments : 0 6 9
    # 7 segments : 8

    map = {}

    one = d[2].pop()
    seven = d[3].pop()
    four = d[4].pop()
    eight = d[7].pop()

    for x in d[6]:  # 0 6 9
        if one[0] in x and one[1] in x:
            # si on n'a pas les deux segments du 1
            if four[0] in x and four[1] in x and four[2] in x and four[3] in x:
                # si 4 est inclus, c'est forcément un 9
                nine = x
            else:
                # sinon, c'est forcément un 0
                zero = x
        else:
            # sinon, c'est forcément un 6
            six = x

    for x in d[5]:  # 2 3 5
        # si 7 est inclus, c est forcément un 3
        if seven[0] in x and seven[1] in x and seven[2] in x:
            three = x
        # si c'est inclus dans 9, c'est forcément un 5
        elif x[0] in nine and x[1] in nine and x[2] in nine and x[3] in nine and x[4] in nine:
            five = x
        # sinon, c'est forcément un 2
        else:
            two = x

    # print(zero, ": 0")
    # print(one, ": 1")
    # print(two, ": 2")
    # print(three, ": 3")
    # print(four, ": 4")
    # print(five, ": 5")
    # print(six, ": 6")
    # print(seven, ": 7")
    # print(eight, ": 8")
    # print(nine, ": 9")

    map[zero] = 0
    map[one] = 1
    map[two] = 2
    map[three] = 3
    map[four] = 4
    map[five] = 5
    map[six] = 6
    map[seven] = 7
    map[eight] = 8
    map[nine] = 9

    r = 0
    for d in code:
        c = map["".join(sorted(d))]
        r = r * 10 + c
    total += r
print(total)
